File: apps/workflow-engine/src/nodes/AIAgent.ts
import type { NodeDefinition } from '../schemas/workflow.js';
import type {
  ExecutionContext,
  NodeData,
  NodeExecutionResult,
} from '../engine/types.js';
import type { INodeTypeDescription } from '../engine/nodeSchema.js';
import { BaseNode } from './BaseNode.js';
import {
  getToolDeclarations,
  executeTool,
  listTools,
  type ToolDeclaration,
} from './ai/tools.js';

/**
 * Gemini API types for function calling
 */
interface GeminiTool {
  functionDeclarations: ToolDeclaration[];
}

interface GeminiContent {
  role: 'user' | 'model' | 'function';
  parts: Array<
    | { text: string }
    | { functionCall: { name: string; args: Record<string, unknown> } }
    | { functionResponse: { name: string; response: Record<string, unknown> } }
  >;
}

interface GeminiRequest {
  contents: GeminiContent[];
  tools?: GeminiTool[];
  systemInstruction?: { parts: Array<{ text: string }> };
  generationConfig?: {
    temperature?: number;
    maxOutputTokens?: number;
  };
}

interface GeminiResponse {
  candidates: Array<{
    content: {
      parts: Array<
        | { text?: string }
        | { functionCall?: { name: string; args: Record<string, unknown> } }
      >;
      role: string;
    };
    finishReason: string;
  }>;
  usageMetadata?: {
    promptTokenCount: number;
    candidatesTokenCount: number;
    totalTokenCount: number;
  };
}

/**
 * AI Agent Node - Agentic LLM with tool calling using Google Gemini
 *
 * This node implements an agentic loop:
 * 1. Send prompt + tools to LLM
 * 2. If LLM requests a tool call, execute it
 * 3. Feed result back to LLM
 * 4. Repeat until LLM is done or max iterations reached
 *
 * Tools are loaded from the shared tools registry (src/nodes/ai/tools.ts)
 */
export class AIAgentNode extends BaseNode {
  readonly type = 'AIAgent';
  readonly description = 'AI Agent with tool calling capabilities using Gemini';

  static readonly nodeDescription: INodeTypeDescription = {
    name: 'AIAgent',
    displayName: 'AI Agent',
    icon: 'fa:robot',
    description: 'AI Agent with tool calling capabilities using Gemini',
    group: ['ai'],
    inputs: [{ name: 'main', displayName: 'Input', type: 'main' }],
    outputs: [
      {
        name: 'main',
        displayName: 'Output',
        type: 'main',
        schema: {
          type: 'object',
          properties: {
            response: { type: 'string', description: 'Agent final response text' },
            toolCalls: {
              type: 'array',
              description: 'List of tool calls made by the agent',
              items: {
                type: 'object',
                properties: {
                  tool: { type: 'string', description: 'Tool name' },
                  args: { type: 'unknown', description: 'Tool arguments' },
                  result: { type: 'unknown', description: 'Tool result' },
                },
              },
            },
            iterations: { type: 'number', description: 'Number of agent loop iterations' },
            model: { type: 'string', description: 'Model used for generation' },
            usage: {
              type: 'object',
              description: 'Token usage statistics',
              properties: {
                promptTokens: { type: 'number', description: 'Total tokens in prompts' },
                completionTokens: { type: 'number', description: 'Total tokens in completions' },
                totalTokens: { type: 'number', description: 'Total tokens used' },
              },
            },
          },
        },
      },
    ],

    properties: [
      {
        displayName: 'API Key',
        name: 'apiKey',
        type: 'string',
        default: '',
        required: true,
        description:
          'Google AI API Key. Get yours at https://aistudio.google.com/app/apikey',
        typeOptions: { password: true },
      },
      {
        displayName: 'Model',
        name: 'model',
        type: 'options',
        default: 'gemini-2.5-flash',
        required: true,
        options: [
          {
            name: 'Gemini 2.5 Flash',
            value: 'gemini-2.5-flash',
            description: 'Fast and efficient, supports 1M tokens',
          },
          {
            name: 'Gemini 2.5 Pro',
            value: 'gemini-2.5-pro',
            description: 'Most capable, best for complex tasks',
          },
          {
            name: 'Gemini 2.0 Flash',
            value: 'gemini-2.0-flash',
            description: 'Fast and versatile multimodal model',
          },
        ],
      },
      {
        displayName: 'System Prompt',
        name: 'systemPrompt',
        type: 'string',
        default:
          'You are a helpful assistant. Use the available tools when needed to complete tasks.',
        description: 'Instructions for how the AI agent should behave',
        typeOptions: { rows: 4 },
      },
      {
        displayName: 'User Prompt',
        name: 'userPrompt',
        type: 'string',
        default: '',
        required: true,
        description:
          'The task for the agent to complete. Supports expressions: {{ $json.task }}',
        typeOptions: { rows: 6 },
      },
      {
        displayName: 'Tools',
        name: 'tools',
        type: 'json',
        default: '["http_request", "calculate", "get_current_time"]',
        description: `JSON array of tools to enable. Available: ${listTools().map((t) => `"${t}"`).join(', ')}`,
        typeOptions: { rows: 3 },
      },
      {
        displayName: 'Max Iterations',
        name: 'maxIterations',
        type: 'number',
        default: 10,
        description:
          'Maximum number of tool call iterations to prevent infinite loops',
        typeOptions: { minValue: 1, maxValue: 50 },
      },
      {
        displayName: 'Temperature',
        name: 'temperature',
        type: 'number',
        default: 0.7,
        description: 'Controls randomness (0 = deterministic, 1 = creative)',
        typeOptions: { minValue: 0, maxValue: 1, step: 0.1 },
      },
      {
        displayName: 'Max Tokens',
        name: 'maxTokens',
        type: 'number',
        default: 2048,
        description: 'Maximum number of tokens in each response',
        typeOptions: { minValue: 1, maxValue: 8192 },
      },
    ],
  };

  async execute(
    _context: ExecutionContext,
    nodeDefinition: NodeDefinition,
    inputData: NodeData[]
  ): Promise<NodeExecutionResult> {
    // Get API key from parameter, fallback to environment variable
    const apiKeyParam = this.getParameter<string>(nodeDefinition, 'apiKey', '');
    const apiKey = apiKeyParam || process.env.GOOGLE_AI_API_KEY;
    if (!apiKey) {
      throw new Error(
        'API Key is required. Enter it in the node settings or set GOOGLE_AI_API_KEY environment variable.'
      );
    }

    const model = this.getParameter<string>(nodeDefinition, 'model');
    const systemPrompt = this.getParameter<string>(
      nodeDefinition,
      'systemPrompt',
      ''
    );
    const userPrompt = this.getParameter<string>(nodeDefinition, 'userPrompt');
    const toolsParam = this.getParameter<string | string[]>(
      nodeDefinition,
      'tools',
      '["http_request", "calculate", "get_current_time"]'
    );
    // Handle both JSON string and array formats
    const selectedTools: string[] =
      typeof toolsParam === 'string' ? JSON.parse(toolsParam) : toolsParam;
    const maxIterations = this.getParameter<number>(
      nodeDefinition,
      'maxIterations',
      10
    );
    const temperature = this.getParameter<number>(
      nodeDefinition,
      'temperature',
      0.7
    );
    const maxTokens = this.getParameter<number>(
      nodeDefinition,
      'maxTokens',
      2048
    );

    const results: NodeData[] = [];

    // Process each input item
    for (const item of inputData.length > 0 ? inputData : [{ json: {} }]) {
      const response = await this.runAgentLoop(apiKey, {
        model,
        systemPrompt,
        userPrompt,
        selectedTools,
        maxIterations,
        temperature,
        maxTokens,
        inputData: item.json,
      });

      results.push({
        json: {
          ...item.json,
          response: response.text,
          toolCalls: response.toolCalls,
          iterations: response.iterations,
          model,
          usage: response.usage,
        },
      });
    }

    return this.output(results);
  }

  private async runAgentLoop(
    apiKey: string,
    options: {
      model: string;
      systemPrompt: string;
      userPrompt: string;
      selectedTools: string[];
      maxIterations: number;
      temperature: number;
      maxTokens: number;
      inputData: Record<string, unknown>;
    }
  ): Promise<{
    text: string;
    toolCalls: Array<{ tool: string; args: unknown; result: unknown }>;
    iterations: number;
    usage: Record<string, number>;
  }> {
    const {
      model,
      systemPrompt,
      userPrompt,
      selectedTools,
      maxIterations,
      temperature,
      maxTokens,
      inputData,
    } = options;

    // Get tool declarations from the shared registry
    const toolDeclarations = getToolDeclarations(selectedTools);
    const tools: GeminiTool | undefined =
      toolDeclarations.length > 0
        ? { functionDeclarations: toolDeclarations }
        : undefined;

    // Initialize conversation
    const contents: GeminiContent[] = [
      {
        role: 'user',
        parts: [
          {
            text: `${userPrompt}\n\nContext data: ${JSON.stringify(inputData)}`,
          },
        ],
      },
    ];

    const toolCalls: Array<{ tool: string; args: unknown; result: unknown }> =
      [];
    let totalUsage = { promptTokens: 0, completionTokens: 0, totalTokens: 0 };
    let iterations = 0;
    let finalText = '';

    // Agent loop
    while (iterations < maxIterations) {
      iterations++;

      const response = await this.callGemini(apiKey, {
        model,
        contents,
        tools,
        systemPrompt,
        temperature,
        maxTokens,
      });

      // Accumulate usage
      if (response.usage) {
        totalUsage.promptTokens += response.usage.promptTokens;
        totalUsage.completionTokens += response.usage.completionTokens;
        totalUsage.totalTokens += response.usage.totalTokens;
      }

      const candidate = response.candidates[0];
      const parts = candidate.content.parts;

      // Check for function calls
      const functionCalls = parts.filter(
        (
          p
        ): p is {
          functionCall: { name: string; args: Record<string, unknown> };
        } => 'functionCall' in p
      );

      if (functionCalls.length > 0) {
        // Add assistant response to conversation
        contents.push({
          role: 'model',
          parts: parts as GeminiContent['parts'],
        });

        // Execute each function call using shared tool executor
        const functionResponses: GeminiContent['parts'] = [];

        for (const fc of functionCalls) {
          const toolName = fc.functionCall.name;
          const toolArgs = fc.functionCall.args;

          const result = await executeTool(toolName, toolArgs);

          if (result.success) {
            toolCalls.push({ tool: toolName, args: toolArgs, result: result.result });
            functionResponses.push({
              functionResponse: {
                name: toolName,
                response: result.result as Record<string, unknown>,
              },
            });
          } else {
            toolCalls.push({
              tool: toolName,
              args: toolArgs,
              result: { error: result.error },
            });
            functionResponses.push({
              functionResponse: {
                name: toolName,
                response: { error: result.error },
              },
            });
          }
        }

        // Add function responses to conversation
        contents.push({
          role: 'function',
          parts: functionResponses,
        });

        // Continue the loop
        continue;
      }

      // No function calls - we have a final response
      const textParts = parts.filter(
        (p): p is { text: string } => 'text' in p && typeof p.text === 'string'
      );
      finalText = textParts.map((p) => p.text).join('');
      break;
    }

    return {
      text: finalText,
      toolCalls,
      iterations,
      usage: totalUsage,
    };
  }

  private async callGemini(
    apiKey: string,
    options: {
      model: string;
      contents: GeminiContent[];
      tools?: GeminiTool;
      systemPrompt: string;
      temperature: number;
      maxTokens: number;
    }
  ): Promise<GeminiResponse & { usage: Record<string, number> }> {
    const { model, contents, tools, systemPrompt, temperature, maxTokens } =
      options;

    const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;

    const requestBody: GeminiRequest = {
      contents,
      generationConfig: {
        temperature,
        maxOutputTokens: maxTokens,
      },
    };

    if (tools) {
      requestBody.tools = [tools];
    }

    if (systemPrompt) {
      requestBody.systemInstruction = {
        parts: [{ text: systemPrompt }],
      };
    }

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Gemini API error (${response.status}): ${error}`);
    }

    const data = (await response.json()) as GeminiResponse;

    if (!data.candidates || data.candidates.length === 0) {
      throw new Error('No response from Gemini API');
    }

    return {
      ...data,
      usage: {
        promptTokens: data.usageMetadata?.promptTokenCount || 0,
        completionTokens: data.usageMetadata?.candidatesTokenCount || 0,
        totalTokens: data.usageMetadata?.totalTokenCount || 0,
      },
    };
  }
}

File: apps/workflow-engine/src/nodes/ai_agent.py
"""AI Agent node - agentic loop with tool calling."""

from __future__ import annotations

import json
from typing import Any, TYPE_CHECKING

from .base import (
    BaseNode,
    NodeTypeDescription,
    NodeInputDefinition,
    NodeOutputDefinition,
    NodeProperty,
    NodePropertyOption,
)
from ..engine.types import SubnodeSlotDefinition

if TYPE_CHECKING:
    from ..engine.types import (
        ExecutionContext,
        NodeData,
        NodeDefinition,
        NodeExecutionResult,
        SubnodeContext,
    )


class AIAgentNode(BaseNode):
    """AI Agent node - agentic loop with tool calling capabilities."""

    node_description = NodeTypeDescription(
        name="AIAgent",
        display_name="AI Agent",
        description="Autonomous agent with tool calling capabilities",
        icon="fa:brain",
        group=["ai"],
        inputs=[NodeInputDefinition(name="main", display_name="Input")],
        outputs=[
            NodeOutputDefinition(
                name="main",
                display_name="Result",
                schema={
                    "type": "object",
                    "properties": {
                        "response": {"type": "string", "description": "Final agent response"},
                        "toolCalls": {"type": "array", "description": "Tools called during execution"},
                        "iterations": {"type": "number", "description": "Number of agent iterations"},
                    },
                },
            )
        ],
        properties=[
            NodeProperty(
                display_name="Model",
                name="model",
                type="options",
                default="gemini-2.0-flash",
                options=[
                    NodePropertyOption(name="Gemini 2.0 Flash", value="gemini-2.0-flash"),
                    NodePropertyOption(name="Gemini 1.5 Flash", value="gemini-1.5-flash"),
                    NodePropertyOption(name="Gemini 1.5 Pro", value="gemini-1.5-pro"),
                    NodePropertyOption(name="GPT-4o", value="gpt-4o"),
                    NodePropertyOption(name="GPT-4o Mini", value="gpt-4o-mini"),
                    NodePropertyOption(name="Claude Sonnet", value="claude-sonnet-4-20250514"),
                ],
            ),
            NodeProperty(
                display_name="System Prompt",
                name="systemPrompt",
                type="string",
                default="You are a helpful AI assistant with access to tools.",
                type_options={"rows": 3},
            ),
            NodeProperty(
                display_name="Task",
                name="task",
                type="string",
                default="",
                required=True,
                description="Task for the agent. Supports expressions.",
                type_options={"rows": 5},
            ),
            NodeProperty(
                display_name="Tools",
                name="tools",
                type="collection",
                default=[],
                type_options={"multipleValues": True},
                properties=[
                    NodeProperty(
                        display_name="Tool Name",
                        name="name",
                        type="string",
                        default="",
                    ),
                    NodeProperty(
                        display_name="Description",
                        name="description",
                        type="string",
                        default="",
                    ),
                    NodeProperty(
                        display_name="Parameters (JSON Schema)",
                        name="parameters",
                        type="json",
                        default="{}",
                    ),
                ],
            ),
            NodeProperty(
                display_name="Max Iterations",
                name="maxIterations",
                type="number",
                default=10,
                description="Maximum number of agent iterations",
            ),
            NodeProperty(
                display_name="Temperature",
                name="temperature",
                type="number",
                default=0.7,
            ),
        ],
        subnode_slots=[
            SubnodeSlotDefinition(
                name="chatModel",
                display_name="Model",
                slot_type="model",
                required=False,
                multiple=False,
            ),
            SubnodeSlotDefinition(
                name="memory",
                display_name="Memory",
                slot_type="memory",
                required=False,
                multiple=False,
            ),
            SubnodeSlotDefinition(
                name="tools",
                display_name="Tools",
                slot_type="tool",
                required=False,
                multiple=True,
            ),
        ],
    )

    @property
    def type(self) -> str:
        return "AIAgent"

    @property
    def description(self) -> str:
        return "Autonomous agent with tool calling capabilities"

    async def execute(
        self,
        context: ExecutionContext,
        node_definition: NodeDefinition,
        input_data: list[NodeData],
        subnode_context: SubnodeContext | None = None,
    ) -> NodeExecutionResult:
        from ..engine.types import NodeData

        # Get parameters with defaults
        model = self.get_parameter(node_definition, "model", "gemini-2.0-flash")
        system_prompt = self.get_parameter(node_definition, "systemPrompt", "")
        task = self.get_parameter(node_definition, "task", "")
        tools_config = self.get_parameter(node_definition, "tools", [])
        max_iterations = self.get_parameter(node_definition, "maxIterations", 10)
        temperature = self.get_parameter(node_definition, "temperature", 0.7)

        # Override with model subnode config if connected
        model_config = self._get_model_config(subnode_context)
        if model_config:
            model = model_config.get("model", model)
            temperature = model_config.get("temperature", temperature)

        # Get memory functions if connected
        memory_config = self._get_memory_config(subnode_context)

        if not task:
            raise ValueError("Task is required")

        # Build tools from connected subnodes first, then from config
        tools, tool_executors = self._build_tools_from_subnodes(subnode_context)

        # Add tools from parameter config if no subnodes connected
        if not tools:
            tools = self._build_tools(tools_config)

        results: list[NodeData] = []

        for item in input_data if input_data else [NodeData(json={})]:
            # Build task with context
            context_str = json.dumps(item.json, indent=2) if item.json else ""

            # Inject chat history from memory if available
            chat_history = ""
            if memory_config and "getHistoryText" in memory_config:
                chat_history = memory_config["getHistoryText"]()

            # Build full task with history and input
            full_task = task
            if chat_history:
                full_task = f"Previous conversation:\n{chat_history}\n\nCurrent request: {task}"
            if context_str:
                full_task = f"{full_task}\n\nInput data:\n{context_str}"

            result = await self._run_agent_loop(
                model=model,
                system_prompt=system_prompt,
                task=full_task,
                tools=tools,
                tool_executors=tool_executors,
                max_iterations=max_iterations,
                temperature=temperature,
            )

            # Save to memory if connected
            if memory_config and "addMessage" in memory_config:
                memory_config["addMessage"]("user", task)
                if result.get("response"):
                    memory_config["addMessage"]("assistant", result["response"])

            results.append(NodeData(json=result))

        return self.output(results)

    def _get_model_config(self, subnode_context: SubnodeContext | None) -> dict[str, Any] | None:
        """Get model configuration from connected model subnode."""
        if not subnode_context or not subnode_context.models:
            return None
        # Use first connected model
        return subnode_context.models[0].config

    def _get_memory_config(self, subnode_context: SubnodeContext | None) -> dict[str, Any] | None:
        """Get memory configuration from connected memory subnode."""
        if not subnode_context or not subnode_context.memory:
            return None
        # Use first connected memory
        return subnode_context.memory[0].config

    def _build_tools_from_subnodes(
        self, subnode_context: SubnodeContext | None
    ) -> tuple[list[dict[str, Any]], dict[str, Any]]:
        """Build tools from connected tool subnodes."""
        tools = []
        tool_executors: dict[str, Any] = {}

        if not subnode_context or not subnode_context.tools:
            return tools, tool_executors

        for resolved_tool in subnode_context.tools:
            config = resolved_tool.config
            if not config.get("name"):
                continue

            tools.append({
                "name": config["name"],
                "description": config.get("description", ""),
                "input_schema": config.get("input_schema", {}),
            })

            # Store executor if provided
            if "execute" in config:
                tool_executors[config["name"]] = config["execute"]

        return tools, tool_executors

    def _build_tools(self, tools_config: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """Build tools array from config."""
        tools = []

        for tool in tools_config:
            if not tool.get("name"):
                continue

            params = tool.get("parameters", {})
            if isinstance(params, str):
                try:
                    params = json.loads(params)
                except json.JSONDecodeError:
                    params = {}

            tools.append({
                "name": tool["name"],
                "description": tool.get("description", ""),
                "input_schema": params,
            })

        # Add built-in tools if no custom tools specified
        if not tools:
            tools = [
                {
                    "name": "search",
                    "description": "Search for information",
                    "input_schema": {
                        "type": "object",
                        "properties": {
                            "query": {"type": "string", "description": "Search query"},
                        },
                        "required": ["query"],
                    },
                },
                {
                    "name": "calculate",
                    "description": "Perform a calculation",
                    "input_schema": {
                        "type": "object",
                        "properties": {
                            "expression": {"type": "string", "description": "Math expression"},
                        },
                        "required": ["expression"],
                    },
                },
            ]

        return tools

    async def _run_agent_loop(
        self,
        model: str,
        system_prompt: str,
        task: str,
        tools: list[dict[str, Any]],
        tool_executors: dict[str, Any],
        max_iterations: int,
        temperature: float,
    ) -> dict[str, Any]:
        """Run an agentic tool-calling loop via the unified LLM provider."""
        from ..engine.llm_provider import get_llm_provider, LLMMessage, LLMTool

        provider = get_llm_provider()

        # Convert tools to LLMTool format
        llm_tools: list[LLMTool] = []
        for t in tools:
            llm_tools.append(LLMTool(
                name=t["name"],
                description=t["description"],
                parameters=t.get("input_schema", {}),
            ))

        # Build initial messages
        messages: list[LLMMessage] = []
        if system_prompt:
            messages.append(LLMMessage(role="system", content=system_prompt))
        messages.append(LLMMessage(role="user", content=task))

        tool_calls_list: list[dict[str, Any]] = []
        iterations = 0

        while iterations < max_iterations:
            iterations += 1

            response = await provider.chat(
                model=model,
                messages=messages,
                tools=llm_tools if llm_tools else None,
                temperature=temperature,
                max_tokens=4096,
            )

            if response.tool_calls:
                # Append assistant message with tool calls
                messages.append(LLMMessage(
                    role="assistant",
                    content=response.content,
                    tool_calls=response.tool_calls,
                ))

                # Execute each tool and append results
                for tc in response.tool_calls:
                    tool_calls_list.append({
                        "tool": tc.name,
                        "input": tc.arguments,
                        "id": tc.id,
                    })

                    tool_result = self._execute_tool(tc.name, tc.arguments, tool_executors)
                    result_str = json.dumps(tool_result) if not isinstance(tool_result, str) else tool_result

                    messages.append(LLMMessage(
                        role="tool",
                        content=result_str,
                        tool_call_id=tc.id,
                    ))

                # Continue loop for model to process tool results
                continue
            else:
                # No tool calls — final text response
                return {
                    "response": response.content or "",
                    "toolCalls": tool_calls_list,
                    "iterations": iterations,
                }

        return {
            "response": "Agent reached maximum iterations",
            "toolCalls": tool_calls_list,
            "iterations": iterations,
        }

    def _execute_tool(
        self, name: str, input_data: dict[str, Any], tool_executors: dict[str, Any]
    ) -> Any:
        """Execute a tool using custom executor or built-in fallback."""
        # Try custom executor first (from connected subnodes)
        if name in tool_executors:
            executor = tool_executors[name]
            try:
                return executor(input_data)
            except Exception as e:
                return {"error": str(e)}

        # Built-in fallback tools
        if name == "search":
            return {"results": f"Search results for: {input_data.get('query', '')}"}
        elif name == "calculate":
            try:
                expr = input_data.get("expression", "0")
                allowed = {"abs": abs, "round": round, "min": min, "max": max, "pow": pow}
                result = eval(expr, {"__builtins__": {}}, allowed)
                return {"result": result}
            except Exception:
                return {"error": "Invalid expression"}
        else:
            return {"message": f"Tool '{name}' executed with input: {input_data}"}
